# WEATHER DATASET EXPLORATION USING PYTHON

# 🌦️ Weather Data Validation

This notebook performs data validation on the weather dataset used in the Coffee Shop BI project.

The goal is to ensure data quality before integration into Power BI.

In [2]:
!pip install pandas
import os
import pandas as pd


[notice] A new release of pip is available: 26.0.1 -> 26.1
[notice] To update, run: C:\Users\keren mayer\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable


## 1. Data Loading

#### Inspecting available sheets in the Excel file

In [4]:

file_path = r"C:\Users\keren mayer\OneDrive\Documents\Bureau\EXPERIS\FINAL BI PROJECT\New York Weather Data.xlsx"

xls = pd.ExcelFile(file_path)
xls.sheet_names

['New York Weather Data']

#### Loading the weather sheet

In [5]:
df_weather = pd.read_excel(file_path, sheet_name="New York Weather Data")
df_weather.head()

,dt,dt_iso,timezone,city_name,lat,lon,temp,visibility,dew_point,feels_like,...,wind_gust,rain_1h,rain_3h,snow_1h,snow_3h,clouds_all,weather_id,weather_main,weather_description,weather_icon
0,1672531200,2023-01-01 00:00:00 +0000 UTC,-18000,New York,40.712775,-74.005973,9.33,2414.0,8.57,7.65,...,0.0,0.25,NaN,NaN,NaN,100,500,Rain,light rain,10n
1,1672531200,2023-01-01 00:00:00 +0000 UTC,-18000,New York,40.712775,-74.005973,9.33,2414.0,8.57,7.65,...,0.0,0.25,NaN,NaN,NaN,100,701,Mist,mist,50n
2,1672534800,2023-01-01 01:00:00 +0000 UTC,-18000,New York,40.712775,-74.005973,9.17,2816.0,8.41,7.77,...,0.0,0.81,NaN,NaN,NaN,100,500,Rain,light rain,10n
3,1672534800,2023-01-01 01:00:00 +0000 UTC,-18000,New York,40.712775,-74.005973,9.17,2816.0,8.41,7.77,...,0.0,0.81,NaN,NaN,NaN,100,701,Mist,mist,50n
4,1672538400,2023-01-01 02:00:00 +0000 UTC,-18000,New York,40.712775,-74.005973,10.05,NaN,9.44,9.62,...,0.0,1.27,NaN,NaN,NaN,100,501,Rain,moderate rain,10n


## 2. Initial Data Exploration

In [6]:
df_weather.shape

(9280, 28)

In [7]:
df_weather.columns

Index(['dt', 'dt_iso', 'timezone', 'city_name', 'lat', 'lon', 'temp',
       'visibility', 'dew_point', 'feels_like', 'temp_min', 'temp_max',
       'pressure', 'sea_level', 'grnd_level', 'humidity', 'wind_speed',
       'wind_deg', 'wind_gust', 'rain_1h', 'rain_3h', 'snow_1h', 'snow_3h',
       'clouds_all', 'weather_id', 'weather_main', 'weather_description',
       'weather_icon'],
      dtype='object')

## 3. Data Structure Analysis

In [8]:
df_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9280 entries, 0 to 9279
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   dt                   9280 non-null   int64  
 1   dt_iso               9280 non-null   object 
 2   timezone             9280 non-null   int64  
 3   city_name            9280 non-null   object 
 4   lat                  9280 non-null   float64
 5   lon                  9280 non-null   float64
 6   temp                 9280 non-null   float64
 7   visibility           9262 non-null   float64
 8   dew_point            9280 non-null   float64
 9   feels_like           9280 non-null   float64
 10  temp_min             9280 non-null   float64
 11  temp_max             9280 non-null   float64
 12  pressure             9280 non-null   int64  
 13  sea_level            0 non-null      float64
 14  grnd_level           0 non-null      float64
 15  humidity             9280 non-null   i

In [11]:
df_weather["dt_iso"].nunique()

8760

## 4. Missing Values Analysis

In [11]:
cols_of_interest = [
    "temp", "temp_min", "temp_max", "feels_like",
    "humidity", "rain_1h", "snow_1h", "clouds_all",
    "weather_main","weather_description"
]

df_weather[cols_of_interest].isnull().sum()


temp                      0
temp_min                  0
temp_max                  0
feels_like                0
humidity                  0
rain_1h                7275
snow_1h                9268
clouds_all                0
weather_main              0
weather_description       0
dtype: int64

### Key Observations

- No missing values detected in core weather variables such as temperature, humidity, and cloud coverage  
- Missing values are present in `rain_1h` and `snow_1h`  

These missing values are expected, as precipitation is only recorded when it occurs.

Overall, missing values do not indicate data quality issues.

## 5. Time Coverage Validation

In [13]:
df_weather["dt_iso"].min()


'2023-01-01 00:00:00 +0000 UTC'

In [9]:
df_weather["dt_iso"].max()

'2023-12-31 23:00:00 +0000 UTC'

## 6. Time Consistency Check

In [10]:
df_weather["dt_iso"].nunique(), df_weather.shape[0]


(8760, 9280)

The dataset contains 8760 unique timestamps and 9280 total records.

This indicates that some timestamps are duplicated, likely due to overlapping measurements or multiple weather records for the same hour.

This does not significantly impact the analysis but should be noted.

## 7. Weather Variables Validation

In [12]:
df_weather[["temp", "temp_min", "temp_max"]].describe()


,temp,temp_min,temp_max
count,9280.000000,9280.000000,9280.000000
mean,14.236820,13.198805,15.265103
std,8.270332,8.217296,8.363293
min,-14.870000,-15.810000,-13.770000
25%,7.417500,6.637500,8.330000
50%,13.770000,12.750000,14.970000
75%,21.400000,20.530000,22.340000
max,34.370000,33.190000,35.680000


### Key Observations

- Temperature values range from approximately -15°C to 35°C  
- These values are realistic for New York weather conditions  
- No extreme anomalies detected  

Overall, temperature data is consistent and reliable.

## 8. Temperature Consistency Check

In [13]:
(df_weather["temp_max"] < df_weather["temp_min"]).sum()

np.int64(0)

No inconsistencies detected between minimum and maximum temperature values.

All records satisfy the condition:
temp_max ≥ temp_min

This confirms internal consistency of temperature data.

## 9. Conclusion

The weather dataset has been successfully validated and is suitable for integration into the BI model.

- Missing values are expected and limited to precipitation-related fields  
- The dataset covers a complete yearly period with consistent hourly observations  
- Some duplicated timestamps exist but do not significantly impact analysis  
- Temperature values fall within realistic ranges  
- No inconsistencies were detected in temperature relationships  

Overall, the dataset is reliable and ready for use in weather-based analysis.